templates/commands/render.go

```diff
@@ -235,24 +276,182 @@ func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
+	}
+}
+
+// A template parser helper to remove the boilerplate of parsing with our
+// desired options.
+func parseGoTmpl(tpl string) (*template.Template, error) {
+	return template.New("").Option("missingkey=error").Parse(tpl) //nolint:wrapcheck
+}
+
+func parseAndExecuteGoTmpl(m model.String, inputs map[string]string) (string, error) {
+	goTmpl, err := parseGoTmpl(m.Val)
+	if err != nil {
+		return "", model.ErrWithPos(m.Pos, `error compiling as go-template: %w`, err) //nolint:wrapcheck
+	}
+
+	var sb strings.Builder
+	if err := goTmpl.Execute(&sb, inputs); err != nil {
+		return "", model.ErrWithPos(m.Pos, "template.Execute() failed: %w", err) //nolint:wrapcheck
+	}
+
+	return sb.String(), nil
+}
+
+// "from" may be a file or directory. "pos" is only used for error messages.
+func copyRecursive(pos *model.ConfigPos, from, to string, rfs renderFS) error {
+	return fs.WalkDir(rfs, from, func(path string, d fs.DirEntry, err error) error { //nolint:wrapcheck
+		if err != nil {
+			return err // There was some filesystem error. Give up.
+		}
+
+		// We don't have to worry about symlinks here because we passed
+		// DisableSymlinks=true to go-getter.
+
+		relToSrc, err := filepath.Rel(from, path)
+		if err != nil {
+			return model.ErrWithPos(pos, "filepath.Rel(%s,%s): %w", from, path, err) //nolint:wrapcheck
+		}
+
+		dst := filepath.Join(to, relToSrc)
+
+		// The spec file may specify a file to copy that's deep in a
+		// directory tree, without naming its parent directory. We can't
+		// rely on WalkDir having traversed the parent directory of $path,
+		// so we must create the target directory if it doesn't exist.
+		dirToCreate := dst
+		if !d.IsDir() {
+			dirToCreate = filepath.Dir(dst)
+		}
+		if err := rfs.MkdirAll(dirToCreate, 0o700); err != nil {
+			return model.ErrWithPos(pos, "MkdirAll(): %w", err) //nolint:wrapcheck
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		buf, err := rfs.ReadFile(path)
+		if err != nil {
+			return model.ErrWithPos(pos, "ReadFile(): %w", err) //nolint:wrapcheck
+		}
+
+		info, err := rfs.Stat(path)
+		if err != nil {
+			return fmt.Errorf("Stat(): %w", err)
+		}
+
+		// TODO: this arguable should be an atomic write so we don't end up with
+		// half-written files on error.
+		//
+		// The permission bits on the output file are copied from the input file;
+		// this preserves the execute bit on executable files.
+		if err := rfs.WriteFile(dst, buf, info.Mode().Perm()); err != nil {
+			return fmt.Errorf("failed writing to scratch file: WriteFile(): %w", err)
+		}
+
+		return nil
+	})
+}
+
+func actionRegexReplace(ctx context.Context, p *model.RegexReplace, sp *stepParams) error {
+	return fmt.Errorf("not implemented")
+}
+
+func actionStringReplace(ctx context.Context, p *model.StringReplace, sp *stepParams) error {
+	return fmt.Errorf("not implemented")
+}
+
+func actionGoTemplate(ctx context.Context, p *model.GoTemplate, sp *stepParams) error {
+	return fmt.Errorf("not implemented")
+}
```

Comment:
nit: might as well creating separate files for them?

Reply:
Done.